File: DataStructures/SinglyLinkedList/singly_linked_list.js
/* both can be used
function Node(val, next) {
    this.val = (val === undefined ? 0 : val);
    this.next = (next === undefined ? null : next);
}
*/

class Node{
    constructor(val){
        // mistake, was using self instead of this
        this.val = val
        this.next = null
    }
}

class SinglyLinkedList{

    constructor(){
        this.head = null,
        this.tail = null,
        this.length = 0
    }

    push(val){
        // push - add at the end of LL
        let newNode = new Node(val)
        if (!this.head){
            this.head = newNode
            this.tail = this.head
        }
        else {
            this.tail.next = newNode
            this.tail = this.tail.next // same as newNode
        }
        this.length++
        return this   // return the whole list in push
    }

    pop(){
        // pop - removes a node from the end of LL
        if (!this.head) return undefined

        let currNode = this.head
        let newTail = currNode

        if (this.length === 1){
            this.head = null
            this.tail = null
        }
        else {
         while(currNode.next){
            newTail = currNode
            currNode = currNode.next
        }
        
        newTail.next = null
        this.tail = newTail
        }
        this.length--
        return currNode
    }

    shift(){
        // shift - remove a node from the beginning of LL
        if(!this.head) return undefined
        let currNode = this.head
        this.head = currNode.next
        this.length--
        return currNode
    }

    unshift(val){
        // shift - add a node at the beginning of LL
        let newNode = new Node(val)
        if(!this.head) {
            this.head = newNode
            this.tail = newNode
        } else {
            // mistake we should use else here other wise for empty LL condition
            // after the if(!this.head) it will come and execute these lines and mess up
            newNode.next = this.head 
            this.head = newNode
        }
        this.length++
        return this
    }

    get(index){
        // get a value at index
        if (!this.head || index >= this.length || index < 0) return -1
        if (index === 0) return this.head
        let count = 0
        let currNode = this.head
        while(count !== index){
            currNode = currNode.next
            count++
        }
        return currNode
    }

    set(index, val){
        // set a value at index
        if (!index || !val) return undefined
        if (!this.head || index >= this.length || index < 0) return false
        let count = 0
        let currNode = this.head
        while(count !== index){
            currNode = currNode.next
            count++
        }
        currNode.val = val
        return currNode    
    }

    insert(index, val){
        // insert a value at index
        if (!this.head) this.unshift(val)
        if(index === this.length - 1) this.push(val)

        let newNode = new Node(val)
        let currNode = this.head
        let prevNode = currNode
        let count = 0
        while(count !== index){
            prevNode = currNode
            currNode = currNode.next
            count++
        }
        newNode.next = currNode
        prevNode.next = newNode
        this.length++
        return true
    }

    remove(index){
        // insert a value at index
        if(index < 0 || index > this.length + 1 || !this.head) return false
        if (index === 0) this.shift()
        if(index === this.length - 1) this.pop()

        let prevNode = this.get(index - 1)
        let removedNode = prevNode.next
        prevNode.next = removedNode.next
        this.length--
        return removedNode
    }
    
}

let list = new SinglyLinkedList()
console.log(list.push(1))
console.log(list.get(0))
list.push(2)
list.push(3)
list.pop()

File: DSA/LinkedLists/singly_linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
    
    def __str__(self):
        return str(self.val)


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    
    def __str__(self):
        return f"SinglyLinkedList(length={self.length})"
    
    def display(self):
        """Display the linked list as a chain"""
        curr = self.head
        res = []
        while curr:
            res.append(str(curr.val))
            curr = curr.next
        print(' -> '.join(res) if res else 'Empty List')

    def push(self, val):
        """push - add at the end of LL"""
        new_node = Node(val)
        if not self.head:
            self.head = new_node
            self.tail = self.head
        else:
            self.tail.next = new_node
            self.tail = self.tail.next  # same as new_node
        self.length += 1
        return self  # return the whole list in push

    def pop(self):
        """pop - removes a node from the end of LL"""
        if not self.head:
            return None

        curr_node = self.head
        new_tail = curr_node

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            while curr_node.next:
                new_tail = curr_node
                curr_node = curr_node.next

            new_tail.next = None
            self.tail = new_tail

        self.length -= 1
        return curr_node

    def shift(self):
        """shift - remove a node from the beginning of LL"""
        if not self.head:
            return None
        curr_node = self.head
        self.head = curr_node.next
        self.length -= 1
        return curr_node

    def unshift(self, val):
        """unshift - add a node at the beginning of LL"""
        new_node = Node(val)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            # important: use else here, otherwise for empty LL condition
            # after the if (not self.head) it will execute these lines and mess up
            new_node.next = self.head
            self.head = new_node
        self.length += 1
        return self

    def get(self, index):
        """get a value at index"""
        if not self.head or index >= self.length or index < 0:
            return -1
        if index == 0:
            return self.head
        count = 0
        curr_node = self.head
        while count != index:
            curr_node = curr_node.next
            count += 1
        return curr_node

    def set(self, index, val):
        """set a value at index"""
        if not index or not val:
            return None
        if not self.head or index >= self.length or index < 0:
            return False
        count = 0
        curr_node = self.head
        while count != index:
            curr_node = curr_node.next
            count += 1
        curr_node.val = val
        return curr_node

    def insert(self, index, val):
        """insert a value at index"""
        if not self.head:
            self.unshift(val)
        if index == self.length - 1:
            self.push(val)

        new_node = Node(val)
        curr_node = self.head
        prev_node = curr_node
        count = 0
        while count != index:
            prev_node = curr_node
            curr_node = curr_node.next
            count += 1
        new_node.next = curr_node
        prev_node.next = new_node
        self.length += 1
        return True

    def remove(self, index):
        """remove a node at index"""
        if index < 0 or index > self.length + 1 or not self.head:
            return False
        if index == 0:
            self.shift()
        if index == self.length - 1:
            self.pop()

        prev_node = self.get(index - 1)
        removed_node = prev_node.next
        prev_node.next = removed_node.next
        self.length -= 1
        return removed_node


# Test code
list_ll = SinglyLinkedList()
list_ll.push(1)
list_ll.push(2)
list_ll.push(3)
print("List after pushing 1, 2, 3:")
list_ll.display()

print(f"\nGet node at index 0: {list_ll.get(0)}")
print(f"List length: {list_ll.length}")

list_ll.pop()
print("\nList after popping:")
list_ll.display()
